tuning/data/data_preprocessing_utils.py

```diff
@@ -0,0 +1,208 @@
+def combine_sequence(input_element: str, output_element: str, eos_token: str = ""):
+    """Combines / concatenates input & output element.
+
+    Args:
+        input_element: str
+            Input component of the combined sequence.
+        output_element: str
+            Output component of the combined sequence.
+        eos_token: str
+            EOS token associated with the tokenizer. \
+            If passed, it will be concatenated at end
+
+    Returns:
+        str
+            Sequence combined with whitespace.
+    """
+    if not input_element.endswith((" ", "\n", "\t")) and not output_element.startswith(
+        (" ", "\n", "\t")
+    ):
+        return input_element + " " + output_element + eos_token
+    return input_element + output_element + eos_token
+
+
+def get_data_collator(
+    packing: bool,
+    response_template: Optional[str],
+    tokenizer: AutoTokenizer,
+    is_traindata_tokenized: bool,
+    max_seq_length: int,
+) -> Callable:
+    """Create and return the the appropriate collator type based on the configuration for packing,
+    response_template, and dataset_text_field.
+
+    Args:
+        packing: bool
+            Whether or not we should apply packing or not.
+        response_template: Optional[str]
+            Response template to be used for formatting by TRL.
+        tokenizer: AutoTokenizer
+            Loaded tokenizer object to be used by the collator.
+        is_traindata_tokenized: bool
+            Whether train Dataset is tokenized or not
+        max_seq_length: int
+            Max sequence length expected
+
+    Returns:
+        Callable
+            Callable collator to be leveraged by the trainer.
+    """
+
+    if not packing:
+        # TODO: near term - how response template ids are parsed out needs to be cleaned.
+        # The [2:] here applies if response template has \n prefix, it is needed to strip \n,
+        # otherwise template is not found. We will create issue to clean this out after we discuss
+        # data formats and collators we will support.
+        if response_template:
+            response_template_ids = tokenizer.encode(
+                response_template, add_special_tokens=False
+            )[2:]
+            return DataCollatorForCompletionOnlyLM(
+                response_template=response_template_ids,
+                tokenizer=tokenizer,
+                ignore_index=configs.IGNORE_INDEX,
+            )
+        # Note that this automatically pads labels with -100
+        # TODO check if this is sufficient for preprocessed
+        if is_traindata_tokenized:
+            return DataCollatorForSeq2Seq(
+                tokenizer=tokenizer, padding=True, max_length=max_seq_length
+            )
+        raise ValueError(
+            "Could not pick a data collator. Please refer to supported data formats"
+        )
+
+
+def custom_data_formatter(element, template, formatted_dataset_field):
```

Comment:
we need to document why this is needed? what not just use chat templates?

Reply:
Again..this is not new functionality in the data preprocessor this is old code being retained. We can try to add some comments here to explain what the functions are doing slowly as we expand on the documetation.

Reply:
yes pls